README.md
It's the code from this link:
http://mypage.iu.edu/~rdbeer/


ReleaseNotes.txt

Version 1.2 5/16/17

Now includes a second example involving a visually-guided agent based on the model
in the following paper (thanks to Matthew Setzler for this implementation):

Beer, R.D. (2003). The dynamics of active categorical perception in an evolved 
model agent. Adaptive Behavior 11(4):209-243. 

================================================================================

Version 1.1.5 5/28/15

The best individual is now updated every generation if the ReEvalFlag is set 
rather than only when the best performance improves (Thanks to Eduardo Izquierdo
for this fix) 

================================================================================

Version 1.1.4 2/13/15

Added brief discussions of mutation variance selection and population sizing for
multithreaded search to the documentation. 

================================================================================

Version 1.1.3 7/2/12

Fixed a bug in the initialization of random states when populations are resized
(Thanks to Marc Dillon for finding this bug)

================================================================================

Version 1.1.2 6/16/09

Fixed a bug where the random states for each individual in a population weren't
being seeded with as wide a range of seeds as they should have been (Thanks to
Seth Frey for finding this bug)

================================================================================

Version 1.1.1 - 3/8/08

Fixed a bug where random states written to streams weren't saving information
related to Gaussian random variables (Thanks to Chad Seys for pointing out this
bug)

================================================================================

Version 1.1 - 2/16/08

There are two major changes in this release, both aimed at substantially
increasing the speed of execution of an evolutionary search. The first change is
isolated to the CTRNN class, while the second impacts several different
components.


FAST SIGMOIDAL FUNCTION

This release includes a new fast sigmoid option which considerably speeds up the
numerical integration of CTRNNs. The fast sigmoid implementation uses a
table-based method with linear interpolation between table entries. This option
can be enabled by uncommenting the "#define FAST_SIGMOID" line in ctrnn.h.

The accuracy of the approximation is controlled by the constants SigTabRange and
SigTabSize.  SigTabRange determines the range over which the approximation
occurs.  For example, the default value of 15.0 means that the range of
fastsigmoid(x) is [-15,15], with a value of 0.0 for x < -15 and a value of 1.0
for x > 15.  The constant SigTabSize determines the number of table entries
used. Note that the table only holds values of sigmoid(x) for 0 <= x <=
SigTabRange, since sigmoid(x) = 1 - sigmoid(-x) for negative values of x.


MULTITHREADED SEARCH

This release also includes the capability for multithreaded evaluation of
populations on multicore/multiprocessor machines. This capability makes use of
the pthread library, so it can only be used on systems for which this library is
available. Threaded evaluation can be enabled by uncommenting the line "#define
THREADED_SEARCH" in search.h. The number of evaluation threads to use is
specified by the THREAD_COUNT symbol in search.h, which defaults to 2.

Once individuals in a population can be evaluated in parallel, it becomes
essential that evaluation is thread-safe. The most important change that this
requires in TSearch is that a separate random state must be maintained for each
individual in the population, so that the search code can guarantee that the
same result will be obtained for a given random seed no matter how many threads
are used. This requirement has several consequences.

First, random states are now encapsulated into instances of the RandomState
class, and all of the functions previously declared in random.h now become
methods of this class.

Second, TSearch now maintains its own random state for randomizing the initial
population, seeding the random states of each individual, selection, mutation,
etc. The seed for this random state can be set using the new SetRandomSeed()
method defined on the TSearch class.

Third, a random state is now passed as a second argument to all user evaluation
functions. Any randomization that the evaluation function does should use this
random state.

Fourth, the RandomizeState() and RandomizeOutputs() methods of the CTRNN class,
as well as the Reset() method of the LeggedAgent class, now accept a new random
state argument, which they use to perform the requested randomization.

Finally, it is the responsibility of a user of TSearch to ensure that any
evaluation function they write is thread-safe (i.e., user evaluation functions
must maintain no global state).

================================================================================


CTRNN.cpp
// ************************************************************
// A class for continuous-time recurrent neural networks
//
// RDB
//  8/94 Created
//  12/98 Optimized integration
//  1/08 Added table-based fast sigmoid w/ linear interpolation
//
// To Do
//   1. Could optimize a bit further by making StepSize
//      a CTRNN parameter and prescaling RTaus and
//      by providing versions of EulerStep and RK4Step
//      that ignore gains, but it's not clear that the
//      few percent speed increase is really worth it.
// ************************************************************

#include "CTRNN.h"
#include "random.h"
#include <stdlib.h>


// A fast sigmoid implementation using a table w/ linear interpolation
#ifdef FAST_SIGMOID
int SigTableInitFlag = 0;
double SigTab[SigTabSize];

void InitSigmoidTable(void)
{
  if (!SigTableInitFlag) {
    double DeltaX = SigTabRange/(SigTabSize-1);
    for (int i = 0; i <= SigTabSize-1; i++)
      SigTab[i] = sigma(i * DeltaX);
    SigTableInitFlag = 1;
  }
}

double fastsigmoid(double x)
{
  if (x >= SigTabRange) return 1.0;
  if (x < 0) return 1.0 - fastsigmoid(-x);
  double id;
  double frac = modf(x*(SigTabSize-1)/SigTabRange, &id);
  int i = (int)id;
  double y1 = SigTab[i], y2 = SigTab[i+1];

  return y1 + (y2 - y1) * frac;
}
#endif

// ****************************
// Constructors and Destructors
// ****************************

// The constructor

CTRNN::CTRNN(int newsize)
{
	SetCircuitSize(newsize);
#ifdef FAST_SIGMOID
  InitSigmoidTable();
#endif
}


// The destructor

CTRNN::~CTRNN()
{
	SetCircuitSize(0);
}


// *********
// Utilities
// *********

// Resize a circuit.

void CTRNN::SetCircuitSize(int newsize)
{
	size = newsize;
	states.SetBounds(1,size);
	states.FillContents(0.0);
	outputs.SetBounds(1,size);
	outputs.FillContents(0.0);
	biases.SetBounds(1,size);
	biases.FillContents(0.0);
	gains.SetBounds(1,size);
	gains.FillContents(1.0);
	taus.SetBounds(1,size);
	taus.FillContents(1.0);
	Rtaus.SetBounds(1,size);
	Rtaus.FillContents(1.0);
	externalinputs.SetBounds(1,size);
	externalinputs.FillContents(0.0);
	weights.SetBounds(1,size,1,size);
	weights.FillContents(0.0);
	TempStates.SetBounds(1,size);
	TempOutputs.SetBounds(1,size);
	k1.SetBounds(1,size);
	k2.SetBounds(1,size);
	k3.SetBounds(1,size);
	k4.SetBounds(1,size);
}


// *******
// Control
// *******

// Randomize the states or outputs of a circuit.

void CTRNN::RandomizeCircuitState(double lb, double ub)
{
	for (int i = 1; i <= size; i++)
        SetNeuronState(i, UniformRandom(lb, ub));
}

void CTRNN::RandomizeCircuitState(double lb, double ub, RandomState &rs)
{
	for (int i = 1; i <= size; i++)
    SetNeuronState(i, rs.UniformRandom(lb, ub));
}

void CTRNN::RandomizeCircuitOutput(double lb, double ub)
{
	for (int i = 1; i <= size; i++)
        SetNeuronOutput(i, UniformRandom(lb, ub));
}

void CTRNN::RandomizeCircuitOutput(double lb, double ub, RandomState &rs)
{
	for (int i = 1; i <= size; i++)
    SetNeuronOutput(i, rs.UniformRandom(lb, ub));
}



// Integrate a circuit one step using Euler integration.

void CTRNN::EulerStep(double stepsize)
{
  // Update the state of all neurons.
  for (int i = 1; i <= size; i++) {
    double input = externalinputs[i];
    for (int j = 1; j <= size; j++)
      input += weights[j][i] * outputs[j];
    states[i] += stepsize * Rtaus[i] * (input - states[i]);
  }
  // Update the outputs of all neurons.
  for (int i = 1; i <= size; i++)
    outputs[i] = sigmoid(gains[i] * (states[i] + biases[i]));
}


// Integrate a circuit one step using 4th-order Runge-Kutta.

void CTRNN::RK4Step(double stepsize)
{
	int i,j;
	double input;

	// The first step.
	for (i = 1; i <= size; i++) {
		input = externalinputs[i];
		for (j = 1; j <= size; j++)
			input += weights[j][i] * outputs[j];
		k1[i] = stepsize * Rtaus[i] * (input - states[i]);
		TempStates[i] = states[i] + 0.5*k1[i];
		TempOutputs[i] = sigmoid(gains[i]*(TempStates[i]+biases[i]));
	}

	// The second step.
	for (i = 1; i <= size; i++) {
		input = externalinputs[i];
		for (j = 1; j <= size; j++)
			input += weights[j][i] * TempOutputs[j];
		k2[i] = stepsize * Rtaus[i] * (input - TempStates[i]);
		TempStates[i] = states[i] + 0.5*k2[i];
	}
	for (i = 1; i <= size; i++)
		TempOutputs[i] = sigmoid(gains[i]*(TempStates[i]+biases[i]));

	// The third step.
	for (i = 1; i <= size; i++) {
		input = externalinputs[i];
		for (j = 1; j <= size; j++)
			input += weights[j][i] * TempOutputs[j];
		k3[i] = stepsize * Rtaus[i] * (input - TempStates[i]);
		TempStates[i] = states[i] + k3[i];
	}
	for (i = 1; i <= size; i++)
		TempOutputs[i] = sigmoid(gains[i]*(TempStates[i]+biases[i]));

	// The fourth step.
	for (i = 1; i <= size; i++) {
		input = externalinputs[i];
		for (j = 1; j <= size; j++)
			input += weights[j][i] * TempOutputs[j];
		k4[i] = stepsize * Rtaus[i] * (input - TempStates[i]);
		states[i] += (1.0/6.0)*k1[i] + (1.0/3.0)*k2[i] + (1.0/3.0)*k3[i] + (1.0/6.0)*k4[i];
		outputs[i] = sigmoid(gains[i]*(states[i]+biases[i]));
	}
}


// Set the biases of the CTRNN to their center-crossing values

void CTRNN::SetCenterCrossing(void)
{
    double InputWeights, ThetaStar;

    for (int i = 1; i <= CircuitSize(); i++) {
        // Sum the input weights to this neuron
        InputWeights = 0;
        for (int j = 1; j <= CircuitSize(); j++)
            InputWeights += ConnectionWeight(j, i);
        // Compute the corresponding ThetaStar
        ThetaStar = -InputWeights/2;
        SetNeuronBias(i, ThetaStar);
    }
}

void CTRNN::PrintModel(void)
{
  for (int i = 1; i <= size; i++) {
    cout << "Neuron  "<< i << ":\n";
    cout << "taus: " << taus[i] << "\n";
    cout << "biases: " << biases[i] << "\n";
    cout << "gains: " << gains[i] << "\n";
    cout << "It's the Weights:" << "\n";
    for (int j = 1; j <= size; j++){
      cout << "Weight (" << i << " " << j << ") = " << weights[i][j] << "\n";
    }
    cout << "\n-----------------------------------------\n";
  }
}
// ****************
// Input and Output
// ****************

#include <iomanip>

ostream& operator<<(ostream& os, CTRNN& c)
{
	// Set the precision
	os << setprecision(32);
	// Write the size
	os << c.size << endl << endl;
	// Write the time constants
	for (int i = 1; i <= c.size; i++)
		os << c.taus[i] << " ";
	os << endl << endl;
	// Write the biases
	for (int i = 1; i <= c.size; i++)
		os << c.biases[i] << " ";
	os << endl << endl;
	// Write the gains
	for (int i = 1; i <= c.size; i++)
		os << c.gains[i] << " ";
	os << endl << endl;
	// Write the weights
	for (int i = 1; i <= c.size; i++) {
		for (int j = 1; j <= c.size; j++)
			os << c.weights[i][j] << " ";
		os << endl;
	}
	// Return the ostream
	return os;
}

istream& operator>>(istream& is, CTRNN& c)
{
	// Read the size
	int size;
	is >> size;
	c.SetCircuitSize(size);
	// Read the time constants
	for (int i = 1; i <= size; i++) {
		is >> c.taus[i];
		c.Rtaus[i] = 1/c.taus[i];
	}
	// Read the biases
	for (int i = 1; i <= size; i++)
		is >> c.biases[i];
	// Read the gains
	for (int i = 1; i <= size; i++)
		is >> c.gains[i];
	// Read the weights
	for (int i = 1; i <= size; i++)
		for (int j = 1; j <= size; j++)
			is >> c.weights[i][j];
	// Return the istream
	return is;
}


VisualObject.h
// ****************************
// A class for visual objects
//
// Matthew Setzler 4/19/17
// ****************************

#include <cmath>

// A visual ray – line segment of form y = mx + b with start coordinates and length

typedef struct {
  double b;
  double m;
  double startX;
  double startY;
  double length;
} Ray;


// The VisualObject class declaration

class VisualObject {
  public:
    // The constructor
    VisualObject(double ix = 0.0, double iy = 275.0, double vy_ = -3.0, double size_ = 30.0) {
      cx = ix; cy = iy;
      vy = vy_;
      size = size_;
    };

    // The deconstructor
    ~VisualObject() {};

    // Updates ray length if an intersection occurs
    // Assumes agent is 'looking up' at object
    virtual void RayIntersection(Ray &ray) {};

    // Accessors
    void SetPositionX(double x) {cx = x;};
    void SetPositionY(double y) {cy = y;};
    double PositionX() {return cx;};
    double PositionY() {return cy;};

    void Step(double StepSize) {
      cy += StepSize * vy;
    };

  protected:
    double cx,cy,vy,size;
};


// Class for horizontal line segment

class Line : public VisualObject {
public:
  // The constructor
  // size --> length of segment
  Line(double ix = 0.0, double iy = 275.0, double vy_ = -3.0, double size_ = 30.0) 
    : VisualObject(ix,iy,vy_,size_) {};

  // The deconstructor
  ~Line() {};

  void RayIntersection(Ray &ray) {
    double x_intersect;
    if (ray.m == INFINITY) {
      x_intersect = ray.startX;
    } else {
      x_intersect = (cy - ray.b)/ray.m;
    }

    // No intersection
    if ((fabs(cx-x_intersect) > size/2.0) || (cy < 0)) return;

    // Intersection, return distance
    double new_length = sqrt(pow(x_intersect-ray.startX, 2.0) + pow(cy-ray.startY, 2.0));
    if (new_length < ray.length) ray.length = new_length;
  };
};


// Class definition for a circle

class Circle : public VisualObject {
public:
  // The constructor
  // size --> diameter
  Circle(double ix = 0.0, double iy = 275.0, double vy_ = -3.0, double size_ = 30.0)
    : VisualObject(ix,iy,vy_,size_) {};

  // The deconstructor
  ~Circle() {};

  void RayIntersection(Ray &ray) {
    // Special case, vertical ray
    if (ray.m == INFINITY) {
      if (fabs(ray.startX-cx) > size/2) return;

      double x_intersect = ray.startX;
      double A = 1;
      double B = -2*cy;
      double C = pow(cy,2)-pow(size/2,2)+pow(x_intersect-cx,2);
      double disc = pow(B,2)-4*A*C;
      double y_intersect = (-B-sqrt(disc))/(2*A); // assuming cy>0 and cy>ray.startY

      double new_length = fabs(y_intersect-ray.startY);
      if (new_length < ray.length) ray.length = new_length;
      return;
    }

    double A = 1 + pow(ray.m, 2);
    double B = 2*(ray.m*(ray.b-cy)-cx);
    double C = pow(cx,2)+pow(ray.b,2)-2*ray.b*cy+pow(cy,2)-pow(size/2,2);
    double disc = pow(B,2)-4*A*C;

    if (disc < 0) return; // No intersection

    // Find points of intersection, return smaller distance (distances are the same for tangent lines)
    double x_intersect1 = (-B+sqrt(disc))/(2*A);
    double y_intersect1 = ray.m*x_intersect1 + ray.b;
    double distance1 = sqrt(pow(x_intersect1-ray.startX,2)+pow(y_intersect1-ray.startY,2));

    double x_intersect2 = (-B-sqrt(disc))/(2*A);
    double y_intersect2 = ray.m*x_intersect2 + ray.b;
    double distance2 = sqrt(pow(x_intersect2-ray.startX,2)+pow(y_intersect2-ray.startY,2));

    double new_length;
    if (distance1 < distance2) {
      new_length = distance1;
    } else {
      new_length = distance2;
    }
    if (new_length < ray.length) ray.length = new_length;
  };
};


SimpleCTRNNMain.cpp
// ***************************************
// A very simple example program for CTRNN
// ***************************************

#include "CTRNN.h"


// Global constants
const double RunDuration = 250;
const double StepSize = 0.01;

// The main program
int main(int argc, char* argv[])
{
    // Set up the circuit
    CTRNN c(2);
    c.SetNeuronBias(1, -2.75);
    c.SetNeuronBias(2, -1.75);
    c.SetConnectionWeight(1, 1, 4.5);
    c.SetConnectionWeight(1, 2, -1);
    c.SetConnectionWeight(2, 1, 1);
    c.SetConnectionWeight(2, 2, 4.5);

    // Run the circuit
    c.RandomizeCircuitState(-0.5,0.5);
    cout << 0.0 << " " << c.NeuronOutput(1) << " " << c.NeuronOutput(2) << endl;
    for (double time = StepSize; time <= RunDuration; time += StepSize) {
        c.EulerStep(StepSize);
        cout << time << " " << c.NeuronOutput(1) << " " << c.NeuronOutput(2) << endl;
    }
    
    // Finished
    return 0;
}


SimpleVisualMain.cpp
#include "random.h"
#include "VisualAgent.h"
#include <iostream>
#include <fstream>
#include <chrono>

// Global constants

// const double StepSize = 0.1;
// const double RunDuration = 200;
const long RandomSeed = 1;


// The main program
void run_for_line(ofstream &outFile, double StepSize, int X1, int Y1, int X2, int Y2){
  VisualAgent Agent;
  Line Object;
  // Circle Object;

  // Load the CTRNN into the agent
  char fname[] = "categorize.ns";
  ifstream ifs;
  ifs.open(fname);
  if (!ifs) {
      cerr << "File not found: " << fname << endl;
      exit(EXIT_FAILURE);
  }
  ifs >> Agent.NervousSystem;

  // Run the agent
  SetRandomSeed(RandomSeed);
  Agent.Reset(0, Y1);
  Agent.SetPositionX(X1);
  Object.SetPositionX(X2);
  Object.SetPositionY(Y2);
  int timer = 0;
  int status = 0;
  auto start = std::chrono::system_clock::now();
  for (double t = 0; Object.PositionY() > BodySize/2; t += StepSize) {
      // cout << "------------------\n";
      // cout << Agent.PositionX() << " " << Agent.PositionY() << " \n";
      // cout << Object.PositionX() << " " << Object.PositionY() << " \n";
      outFile << 1 << ',' << ++timer << ',' << StepSize << ',' << X1 << ',' << Y1 << ',' << X2 << ',' << Y2 << ',';
      outFile << Agent.PositionX() << "," << Agent.PositionY() << "," << Object.PositionX() << "," << Object.PositionY() << "," << status <<" \n";
      status = 1;
      Agent.Step(StepSize, Object);
      Object.Step(StepSize);
  }
  auto end = std::chrono::system_clock::now();
  std::chrono::duration<double> elapsed_seconds = end-start;
    std::time_t end_time = std::chrono::system_clock::to_time_t(end);
  cout << "finished computation at " << std::ctime(&end_time)
              << "elapsed time: " << elapsed_seconds.count() << "s\n";

  outFile << 1 << ',' << ++timer << ',' << StepSize << ',' << X1 << ',' << Y1 << ',' << X2 << ',' << Y2 << ',';
  outFile << Agent.PositionX() << "," << Agent.PositionY() << "," << Object.PositionX() << "," << Object.PositionY() << "," << ++status <<" \n";


}

void run_for_Circle(ofstream &outFile, double StepSize, int X1, int Y1, int X2, int Y2){
  VisualAgent Agent;
  // Line Object;
  Circle Object;

  // Load the CTRNN into the agent
  char fname[] = "categorize.ns";
  ifstream ifs;
  ifs.open(fname);
  if (!ifs) {
      cerr << "File not found: " << fname << endl;
      exit(EXIT_FAILURE);
  }
  ifs >> Agent.NervousSystem;

  // Run the agent
  SetRandomSeed(RandomSeed);
  Agent.Reset(0, Y1);
  Agent.SetPositionX(X1);
  Object.SetPositionX(X2);
  Object.SetPositionY(Y2);
  int timer = 0;
  int status = 0;
  for (double t = 0; Object.PositionY() > BodySize/2; t += StepSize) {
      // cout << "------------------\n";
      // cout << Agent.PositionX() << " " << Agent.PositionY() << " \n";
      // cout << Object.PositionX() << " " << Object.PositionY() << " \n";
      outFile << 2 << ',' << ++timer << ',' << StepSize << ',' << X1 << ',' << Y1 << ',' << X2 << ',' << Y2 << ',';
      outFile << Agent.PositionX() << "," << Agent.PositionY() << "," << Object.PositionX() << "," << Object.PositionY() << "," << status <<" \n";
      status = 1;
      Agent.Step(StepSize, Object);
      Object.Step(StepSize);
  }
  outFile << 2 << ',' << ++timer << ',' << StepSize << ',' << X1 << ',' << Y1 << ',' << X2 << ',' << Y2 << ',';
  outFile << Agent.PositionX() << "," << Agent.PositionY() << "," << Object.PositionX() << "," << Object.PositionY() << "," << ++status <<" \n";

}

void showModel(string fname){
  CTRNN NervousSystem;
  ifstream ifs;
  ifs.open(fname);
  if (!ifs) {
      cerr << "File not found: " << fname << endl;
      exit(EXIT_FAILURE);
  }
  ifs >> NervousSystem;
  cout << "------------------------------------\n";
  NervousSystem.PrintModel();
}

int main(int argc, char* argv[])
{
    ofstream outFile;
    outFile.open ("output.csv");
    outFile << "obj_type,timer,step_size,X1,Y1,X2,Y2,agent_X,agent_Y,obj_X,obj_Y,status\n";
    // int X1, Y1, X2, Y2;
    for (int j1=150; j1 < 275; j1+=20){
      for (int j2=-30; j2 <= 30; j2+=15){
        for (double i = 0.1; i <= 0.2; i+= 0.05){
          cout << "------------------------------------\n";
          cout << i << " " << 0<< " " << j1<< " " <<-32 + int((275-j1)/2)<< " " << j2 << "\n";
          run_for_line(outFile, i, j2, 0, -32 + int((275-j1)/2), j1);
          run_for_Circle(outFile, i, j2, 0, -32 + int((275-j1)/2), j1);
        }
      }
    }
    outFile.close();
    showModel("categorize.ns");
    return 0;
}


CTRNN.h
// ***********************************************************
// A class for continuous-time recurrent neural networks
//
// RDB
//  8/94 Created
//  12/98 Optimized integration
//  1/08 Added table-based fast sigmoid w/ linear interpolation
// ************************************************************

// Uncomment the following line for table-based fast sigmoid w/ linear interpolation
//#define FAST_SIGMOID

#include "VectorMatrix.h"
#include "random.h"
#include <iostream>
#include <math.h>

#pragma once


// The sigmoid function

#ifdef FAST_SIGMOID
const int SigTabSize = 400;
const double SigTabRange = 15.0;

double fastsigmoid(double x);
#endif

inline double sigma(double x)
{
  return 1/(1 + exp(-x));
}

inline double sigmoid(double x)
{
#ifndef FAST_SIGMOID
  return sigma(x);
#else
  return fastsigmoid(x);
#endif
}


// The inverse sigmoid function

inline double InverseSigmoid(double y)
{
  return log(y/(1-y));
}


// The CTRNN class declaration

class CTRNN {
    public:
        // The constructor
        CTRNN(int newsize = 0);
        // The destructor
        ~CTRNN();

        // Show the Cells
        void PrintModel(void);

        // Accessors
        int CircuitSize(void) {return size;};
        void SetCircuitSize(int newsize);
        double NeuronState(int i) {return states[i];};
        double &NeuronStateReference(int i) {return states[i];};
        void SetNeuronState(int i, double value)
            {states[i] = value;outputs[i] = sigmoid(gains[i]*(states[i] + biases[i]));};
        double NeuronOutput(int i) {return outputs[i];};
        double &NeuronOutputReference(int i) {return outputs[i];};
        void SetNeuronOutput(int i, double value)
            {outputs[i] = value; states[i] = InverseSigmoid(value)/gains[i] - biases[i];};
        double NeuronBias(int i) {return biases[i];};
        void SetNeuronBias(int i, double value) {biases[i] = value;};
        double NeuronGain(int i) {return gains[i];};
        void SetNeuronGain(int i, double value) {gains[i] = value;};
        double NeuronTimeConstant(int i) {return taus[i];};
        void SetNeuronTimeConstant(int i, double value) {taus[i] = value;Rtaus[i] = 1/value;};
        double NeuronExternalInput(int i) {return externalinputs[i];};
        double &NeuronExternalInputReference(int i) {return externalinputs[i];};
        void SetNeuronExternalInput(int i, double value) {externalinputs[i] = value;};
        double ConnectionWeight(int from, int to) {return weights[from][to];};
        void SetConnectionWeight(int from, int to, double value) {weights[from][to] = value;};
        void LesionNeuron(int n)
        {
            for (int i = 1; i<= size; i++) {
                SetConnectionWeight(i,n,0);
                SetConnectionWeight(n,i,0);
            }
        }
        void SetCenterCrossing(void);

        // Input and output
        friend ostream& operator<<(ostream& os, CTRNN& c);
        friend istream& operator>>(istream& is, CTRNN& c);

        // Control
        void RandomizeCircuitState(double lb, double ub);
        void RandomizeCircuitState(double lb, double ub, RandomState &rs);
        void RandomizeCircuitOutput(double lb, double ub);
        void RandomizeCircuitOutput(double lb, double ub, RandomState &rs);
        void EulerStep(double stepsize);
        void RK4Step(double stepsize);

        int size;
        TVector<double> states, outputs, biases, gains, taus, Rtaus, externalinputs;
        TMatrix<double> weights;
        TVector<double> TempStates,TempOutputs,k1,k2,k3,k4;


};
